collection.any() produces an error if the collection has only one element.

Caused by: org.uqbar.project.wollok.interpreter.WollokInterpreterException: Error evaluating line file:/home/npasserini/checkout/uqbar/wollok/wollok-tests/src/wollok/examples/monstersinc/modelo.wlk:[15]: puertas.any()
		at org.uqbar.project.wollok.interpreter.WollokInterpreter.eval(WollokInterpreter.java:203)
		at org.uqbar.project.wollok.interpreter.WollokInterpreterEvaluator.eval(WollokInterpreterEvaluator.java:131)
		at org.uqbar.project.wollok.interpreter.WollokInterpreterEvaluator$2.apply(WollokInterpreterEvaluator.java:146)
		at org.uqbar.project.wollok.interpreter.WollokInterpreterEvaluator$2.apply(WollokInterpreterEvaluator.java:1)
		at org.eclipse.xtext.xbase.lib.internal.FunctionDelegate.apply(FunctionDelegate.java:42)
		at com.google.common.collect.Lists$TransformingRandomAccessList.get(Lists.java:572)
		at java.util.AbstractList$Itr.next(AbstractList.java:358)
		at java.util.AbstractCollection.toArray(AbstractCollection.java:196)
		at com.google.common.collect.Iterables.toArray(Iterables.java:314)
		at org.eclipse.xtext.xbase.lib.Conversions.unwrapArray(Conversions.java:96)
		at org.uqbar.project.wollok.interpreter.WollokInterpreterEvaluator.evalEach(WollokInterpreterEvaluator.java:149)
		at org.uqbar.project.wollok.interpreter.WollokInterpreterEvaluator._evaluate(WollokInterpreterEvaluator.java:690)
		at org.uqbar.project.wollok.launch.WollokLauncherInterpreterEvaluator.evaluate(WollokLauncherInterpreterEvaluator.java:157)
		at org.uqbar.project.wollok.interpreter.WollokInterpreter.eval(WollokInterpreter.java:191)
		... 69 more
	Caused by: java.lang.IllegalArgumentException: bound must be positive
		at java.util.Random.nextInt(Random.java:388)

--- a/org.uqbar.project.wollok.tests/src/org/uqbar/project/wollok/tests/interpreter/ListTestCase.xtend
+++ b/org.uqbar.project.wollok.tests/src/org/uqbar/project/wollok/tests/interpreter/ListTestCase.xtend
@@ -2,6 +2,7 @@ package org.uqbar.project.wollok.tests.interpreter
 
 import org.uqbar.project.wollok.tests.interpreter.AbstractWollokInterpreterTestCase
 import org.junit.Test
+import org.uqbar.project.wollok.ui.utils.XTendUtilExtensions
 
 /**
  * @author jfernandes
@@ -110,5 +111,34 @@ class ListTestCase extends AbstractWollokInterpreterTestCase {
 			assert.that(greaterThanOneElements.size() == 2)
 		}'''.interpretPropagatingErrors
 	}
-
+	
+	@Test
+	def void testAny() {
+		'''
+		program p {
+			val numbers = #[23]		
+			assert.equals(23, numbers.any())
+		}'''.interpretPropagatingErrors
+	}
+	
+	@Test
+	def void testEqualsWithMethodName() {
+		'''
+		program p {
+			val a = #[23, 2, 1]
+			val b = #[23, 2, 1]
+			assert.that(a.equals(b))
+		}'''.interpretPropagatingErrors
+	}
+	
+	@Test
+	def void testEqualsWithEqualsEquals() {
+		'''
+		program p {
+			val a = #[23, 2, 1]
+			val b = #[23, 2, 1]
+			assert.that(a == b)
+		}'''.interpretPropagatingErrors
+	}
+	
 }

--- a/org.uqbar.project.wollok/src/org/uqbar/project/wollok/interpreter/nativeobj/AbstractWollokDeclarativeNativeObject.xtend
+++ b/org.uqbar.project.wollok/src/org/uqbar/project/wollok/interpreter/nativeobj/AbstractWollokDeclarativeNativeObject.xtend
@@ -18,7 +18,7 @@ abstract class AbstractWollokDeclarativeNativeObject implements WCallable {
 	WollokInterpreterAccess interpreterAccess = new WollokInterpreterAccess
 
 	override call(String message, Object... parameters) {
-		val method = getMethod(message, parameters)
+		val method = getMethod(toJavaMethod(message), parameters)
 		if (method == null)
 			doesNotUnderstand(message, parameters).asWollokObject
 		else
@@ -31,6 +31,12 @@ abstract class AbstractWollokDeclarativeNativeObject implements WCallable {
 				e.printStackTrace
 				throw e
 			}
+	}	
+	
+	def String toJavaMethod(String messageName) {
+		if (messageName == "==") "equals"
+		else
+			messageName
 	}
 
 	def getMethod(String message, Object... parameters) {

--- a/org.uqbar.project.wollok/src/org/uqbar/project/wollok/interpreter/nativeobj/collections/AbstractWollokCollection.xtend
+++ b/org.uqbar.project.wollok/src/org/uqbar/project/wollok/interpreter/nativeobj/collections/AbstractWollokCollection.xtend
@@ -28,6 +28,14 @@ abstract class AbstractWollokCollection<T extends Collection> extends AbstractWo
 	
 	def abstract AbstractWollokCollection<T> asThisCollection(Iterable toWrap)
 	
+	override equals(Object other) { 
+		if (!(other instanceof AbstractWollokCollection)) {
+			return false
+		}
+		val o = other as AbstractWollokCollection
+		size == o.size && this.wrapped.forall[e | o.contains(e) ]
+	}
+	
 	def forEach(WollokClosure proc) { wrapped.forEach(proc.asProc) }
 	def map(WollokClosure closure) { wrapped.map(closure.asFun).asThisCollection }
 	def forAll(WollokClosure pred) { wrapped.forall(pred.asFun) }
@@ -35,6 +43,8 @@ abstract class AbstractWollokCollection<T extends Collection> extends AbstractWo
 	def filter(WollokClosure pred) { wrapped.filter(pred.asFun).asThisCollection }
 	def detect(WollokClosure pred) { wrapped.findFirst(pred.asFun) }
 	def count(WollokClosure pred) { wrapped.filter(pred.asFun).size }
+	
+	def size() { wrapped.size }
 	
 	def join() { join(',') }
 	def join(String separator) { wrapped.join(separator) }
@@ -67,7 +77,7 @@ abstract class AbstractWollokCollection<T extends Collection> extends AbstractWo
 	
 	def Object any() {
 		if (wrapped.isEmpty) throw new WollokRuntimeException("Illegal operation 'any' on empty collection")
-		else wrapped.get(randomBetween(0, wrapped.size - 1))
+		else wrapped.get(randomBetween(0, wrapped.size))
 	}
 	
 	def abstract String getWollokName()

